packages/teleport/src/lib/tdp/client.ts

```diff
@@ -30,6 +32,7 @@ export default class Client extends EventEmitter {
   logger = Logger.create('TDPClient');
   private connected = false;
   private disconnected = false;
+  private errored = false;
```

Comment:
Lets do another round by revisiting this flags as their usage seems unnecessary complicated? For example, do we need this `errored` flag at all? Can we remove `connected` flag?  

```ts
private connected = false;
private disconnected = false;
private errored = false
```  

You may also consider adding a new message type to EventEmiiter 'renderError` - this can help you to distinguish protocol errors vs external errors.

Yeah, lets try to remove all above flags to see if we can do without them.

Reply:
The problem that `disconnected` and `errored` are trying to solve is that when the websocket it closed, the `onclose` handler needs some way to know whether or not to emit the generic `'Connection closed due to an unknown error'` error event. This state must be tracked somehow in order for this object to work how we need it to. 

It would be possible to get rid of the `errored` flag by tracking it with the `CloseEvent.code`, however this would just push the responsibility for setting this aspect of this `Client`'s state to the backend, where we would need to set the websocket's close code depending on whether or not we sent a TDP error. This is a worse design, because now we've spread this event emitter's state to the server, plus it's practically infeasible since we would need to completely refactor the proxy from blindly proxying messages to the websocket to actually decoding TDP and checking for TDP errors, and deciding a websocket code based on them.

Managing this state with websocket close codes raises a similar problem for `disconnect`, though in its case we would need to add a "Disconnect" tdp message, which the proxy would have to be refactored to read and then pass a particular close code over the websocket. 

I agree it would generally be preferable that we could easily just pass these states into `onclose`, however because of the nature of how that listener get's called its just impractical and would lead to a lot of additional complexity in the protocol and on the backend for very little gain. Of course let me know if there's some approach that I'm missing.

It may be possible to get rid of the "connect" event entirely, I'll take a closer look.

Reply:
> the onclose handler needs some way to know whether or not to emit the generic 'Connection closed due to an unknown error' error event. This state must be tracked somehow in order for this object to work how we need it to.

These are 2 types of errors and could be handled separately. 
1) Show TDP error messages when they arrive
2) Show "connection has been terminated" error.

No need to combine them.

Also, lets rename `'Connection closed due to an unknown error'` -> `Connection has been terminated by the server`  as there is no way for us to report an actual error via websocket close event.

You should be able to determine when connection has been closed by the client vs the server.